### DISCUSSION TITLE
Is there a reason why custom task runners are being dropped in v21?

### DISCUSSION THREAD

**Discussion by terrymun** [nrwl/nx#28332] [Questions]
Many of us depend on custom task runners: is there a reason why it will no longer be supported in v21?

![Screenshot 2024-10-07 at 14 28 00](https://github.com/user-attachments/assets/103af3c7-ce1b-42d4-afe2-3977550b34e5)

For example, we are using S3 cache to store our nx distributed cache via [nx-aws-cache](https://www.npmjs.com/package/@nx-aws-plugin/nx-aws-cache), as the Nx Cloud enterprise plan is exorbitantly expensive and offers features that we do not need. All we want is a place to store our distributed cache.

By removing support for custom runners, we will be forced to opt into a paid plan on either Nx Cloud or Nx Powerpack. This kind of vendor-locking is counterproductive.

**Comment by el-davo** [comment]
Agree on this. We use our own on prem cache to. Unfortunately if `tasksRunnerOptions` is removed we will have to look at other options other than NX. Turborepo is looking pretty good right now. NX will have to update this documentation to make it clear that on prem caching is now behind a paywall - https://nx.dev/concepts/turbo-and-nx.

Why would you think it is acceptable to make us pay for storage that we provide and administer ourselves and at such extortionate pricing? 

I think this will prove to be a very bad move in the long run

**Reply by itspers** [reply]
I think they clearly see success of Vercel. Problem is that in general Nrwl was aimed for big corporations and this $249/month is reasonable price for them. Their seniour-lead-staft-CTO-release-production-managers will explain on another meeting that they really need that awesome feature with cloud running agents to distribute that 10.000 tests they wrote because book said - and everybody will be happy.

**Comment by itspers** [comment]
Same for me - i dont need all this Agents and Cloud logging at all. I even dont need CI, my github runs are just building apps and updating them on server. And this builds just fine on included github minutes. All i need is to cache results, so it will not re-build them again for dev and live. And even on this simple usecase i received notification that im reaching some limit on their cloud which i even dont open...

For last years NX has zero useful features for me, only constant problems after each update, instead bombarding with this cloud crap promotions. 

And now only option is to disable this cloud connection at all when i will reach limit and lose caching. It would be definitelly strange to pay extra 26$ per month to have custom caching, when whole github with all this runners costs $4 per user...

> No, we’re not placing existing features behind a paywall. Nx Powerpack introduces new features on top of Nx (more about that below).

https://nx.dev/blog/introducing-nx-powerpack

**Reply by terrymun** [reply]
I find that this particular statement in their article is misleading and disingenuous:

> No, we’re not placing existing features behind a paywall. Nx Powerpack introduces new features on top of Nx (more about that below).

- Disabling custom task runners will mean we cannot use existing task runners to store our Nx distributed cache in a storage solution of our choice
- We will be **forced** to use Nx powerpack, which is a paid feature (by seat)

If the Nx team is dedicated towards OSS, that will mean allowing custom task runners to co-exist with Nx powerpack, but this does not seem to be the case.

**Comment by terrymun** [comment]
Linking to a similar discussion here, which also lists other use cases of Nx custom task runner that will be broken post v21: https://github.com/nrwl/nx/discussions/28360

**Reply by el-davo** [reply]
Also linking this for anyone using this plugin - https://github.com/robinpellegrims/pellegrims/issues/281

**Comment by murilo-arruda** [comment]
I fully agree with the comments above. Removing support for custom task runners is a decision that effectively puts a paywall on a feature many of us rely on to keep our workflows efficient and cost-effective. We’re not asking for extra features or full access to Nx Cloud— all we want is the ability to use a distributed cache solution that we manage and maintain ourselves.

This change puts Nx in a position of vendor lock-in, forcing organizations to subscribe to paid plans for something as basic as cache storage. It undermines the flexibility that many chose Nx for in the first place and alienates smaller teams and companies that simply don’t need all the extra functionality offered by Nx Powerpack or Nx Cloud Enterprise.

If the intention is to make Nx more accessible and keep the OSS community active, support for custom runners should continue to coexist with the new paid options. I hope the Nx team considers the impact this will have on the community and reconsiders this decision before the v21 release.

**Comment by vilkatis** [comment]
This is a very unexpected and disingenuous move if custom task runners are being removed completely 

It basically forces everyone to use one of their two products, when open source solutions were already in place to provide a solution for the teams that don't need or can afford the full functionality of NX Cloud

**Comment by dprothero** [comment]
This is Enshitification at its finest.

**Comment by Pouja** [comment]
What stops them from putting other features in the future behind a paywall as well?

**Comment by PayBas** [comment]
LibreNx here we come!

**Comment by qbitroot** [comment]
If it happens, we will get yet another open source fork drama - a lose-lose situation.

**Reply by haskelcurry** [reply]
It _already_ happens.

**Reply by Hommis** [reply]
Yeah, I think it is time to go with a fork unless the remote caching becomes back again.

**Comment by o629625** [comment]
Suggest the best path forward is keep Powerpack for a premium/white glove/Nx supported remote cache solution. And also keep custom runners for a self serve/open source remote cache solution. There are too many different use cases for custom runners, that Powerpack does not cover.

**Reply by landon-buttars-wgu** [reply]
Funny you mention that. I came across this error when running a change log hook that uses a custom runner. It's a use case that will not be covered by Powerpack.

**Comment by landon-buttars-wgu** [comment]
If Nx goes through with replacing an existing critical feature with a premium paid one will absolutely destroy trust within the organization I work for and there would be no way we would be comfortable using Nx moving forward. DO NOT DO THIS!

**Reply by mjmarianetti** [reply]
same here, if they go this route we will likely move to some other tool. This is the same thing serverless framework did...not great

**Comment by vsavkin** [comment]
Folks, 

We are introducing a new API: https://github.com/nrwl/nx/discussions/29637

It handles most of the functionality that custom tasks runners were used for, but in a much clearer way and without impending our ability to evolve Nx. 

Please take a look.

**Reply by itspers** [reply]
https://github.com/nrwl/nx/discussions/29637#discussioncomment-11848373

Why you guys need to communicate as you are some giant corporation? You could just write few months ago as it is "Sorry people but whole this project does not produce enough money, so we need to extract some logic into paid features" and everybody would understand.

**Reply by IKatsuba** [reply]
@vsavkin the only reason why we use custom runners is remote cache. How does a new API cover that?

**Reply by haskelcurry** [reply]
I agree with previous comments here. I guess it's time to start looking for other options for monorepo (plenty of them).

**Reply by Hommis** [reply]
Will this cover remote caching? If it does not make using custom remote caches possible, there is no other way than to fork nx.

**Reply by terrymun** [reply]
We are already considering offboarding and switching over to Turborepo if remote caching becomes a paid feature on Nx.

**Comment by wuarmin** [comment]
Hey,
can somebody please help me. I just upgraded to latest nx in my pnpm-monorepo. I use nx just as task-runner and use the caching-feature

Here's my `nx.json`. 

```json
{
  "affected": {
    "defaultBase": "main"
  },
  "npmScope": "@my-scope",
  "tasksRunnerOptions": {
    "default": {
      "runner": "nx/tasks-runners/default"
    }
  },
  "targetDefaults": {
    "build": {
      "cache": true,
      "dependsOn": [
        "^build"
      ]
    },
    "start": {
      "dependsOn": [
        "^build"
      ]
    }
  },
  "namedInputs": {
    "default": [
      "{projectRoot}/**/*",
      "sharedGlobals"
    ],
    "sharedGlobals": [
      "{workspaceRoot}/pnpm-workspace.yaml",
      "{workspaceRoot}/pnpm-lock.yaml",
      "{workspaceRoot}/nx.json"
    ],
    "production": [
      "default"
    ]
  }
}
```

Now I get:
```txt
Custom task runners will no longer be supported in Nx 21.
```
if I run i.e. `pnpm nx build my-frontend`. I don't quite understand what the customTaskRunner-options have to do with caching. If I do not want to move to a cloud-cache, I can use `@nx/powerpack-shared-fs-cache` as described here:
https://nx.dev/deprecated/legacy-cache

Am I wrong? If I remove the `tasksRunnerOptions` from my `nx.json`, the deprecation warning is not shown and the caching works as expected.  

For my simple setting, it makes no sense at all to rely on a cloud solution. I am a single user. But I need the caching feature.
What should I be prepared for in version 21?

Thanks

**Comment by vsavkin** [comment]
Folks,

 We changed our plans. **We’re implementing an API to allow self-hosted caching. Everything will be in place before Nx 21, ensuring a smooth transition for everyone. Custom self-hosted caching will possible in all versions of Nx.** 

 Please read this blog post explaining what happened, our new solution, and our OSS commitment to the Nx community:
 [Custom Task Runners and Self-Hosted Caching Changes](https://nx.dev/blog/custom-runners-and-self-hosted-caching)

Given the change, I’m going to close this discussion. If you have any questions or notes regarding the new RFC, please open a discussion.

Thank you so much for engaging with us!
Victor

**Reply by terrymun** [reply]
Thanks for the transparency and the clarity: it is very much appreciated. We remain grateful to be able to keep nx to continue crafting an excellent Dx for our developers across multiple digital teams 👏
